Don't recurse in local backend's List if not required

greatroar opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

The local backend's List method would recursively read subdirectories even for layouts that do not use subdirectories. Due to the return if !isFile, the IsDir branch was never taken.


Also replaced isFile by an IsRegular check, which has been equivalent since Go 1.12 (golang/go@a2a3dd00c934fa15ad880ee5fe1f64308cbc73a7).

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

No.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

Commit:
Don't recurse in local backend's List if not required

Due to the return if !isFile, the IsDir branch in List was never taken
and subdirectories were traversed recursively.

Also replaced isFile by an IsRegular check, which has been equivalent
since Go 1.12 (golang/go@a2a3dd00c934fa15ad880ee5fe1f64308cbc73a7).

MichaelEischer reviewed (commented):
I've noticed that restic picks up files in the root directory even if the layout for that filetype uses subdirectories. Maybe we should change `List()` to skip these files?
For example move `data/ab/abcdef...` to `data/abcdef...`  and watch the error message.

Commit:
backend/local: replace fs.Walk with custom walker

This code is more strict in what it expects to find in the backend:
depending on the layout, either a directory full of files or a directory
full of such directories.

greatroar commented:
I've replaced the fs.Walk with custom directory traversal that breaks when there are files or directories in the wrong place.

MichaelEischer reviewed (approved):
LGTM. Thanks!
